### DISCUSSION TITLE
project metrics (resource quota details)

### DISCUSSION THREAD

**Discussion by dackmail** [openshift/openshift-client-python#99] [Q&A]
is it possible to get project metrics (resource quota information) CPU / memory limits / requests?.
thanks

**Comment by bradmwilliams** [comment]
@dackmail You can use the `oc_action()` method to essentially call any command that is supported by `oc`...
There is an example here: [oc_action.py](https://github.com/openshift/openshift-client-python/blob/main/examples/oc_action.py)
So, something like:
> oc.oc_action(oc.cur_context(), "get", cmd_args=["resourcequotas", "-A"])

Should get you all the resourcequotas on the cluster.

**Reply by dackmail** [reply]
... and to get warning events of projects 
thanks too

**Reply by bradmwilliams** [reply]
> ... and to get warning events of projects

To get all the Warning events on the cluster:
`oc.oc_action(oc.cur_context(), "get", cmd_args=["events", "-A", "--field-selector", "type==Warning"])`

For a specific namespace, you can do either:
`oc.oc_action(oc.cur_context(), "get", cmd_args=["events", "-n", "my-project", "--field-selector", "type==Warning"])`
OR
```
with oc.project('my-project'):
     oc.oc_action(oc.cur_context(), "get", cmd_args=["events", "--field-selector", "type==Warning"])
```

**Reply by dackmail** [reply]
I am new in python :) and  have this message on last code:
<openshift.action.Action object at 0x000001C4c24B5640>

**Reply by bradmwilliams** [reply]
What you are seeing there is the address, in memory, of the  "Action" object itself.  You'll want to do something like this instead:
```
action = oc.oc_action(oc.cur_context(), "get", cmd_args=["events", "--field-selector", "type==Warning"])
print(action.out)
```          

Personally, I would recommend starting with something like this and improving as necessary...
```
#!/usr/bin/python

import argparse
import json
import logging
import traceback
import openshift as oc

from openshift import OpenShiftPythonException

logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
logger = logging.getLogger('oc-tester')
logger.setLevel(logging.INFO)


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Test harness for all things openshift-client-python')
    parser.add_argument('-b', '--bastion', help='The hostname or IP on which to run oc (oc is run locally if not specified)', required=False, default=None)
    parser.add_argument('-v', '--verbose', help='Enable verbose output.', action='store_true')

    args = vars(parser.parse_args())

    if args['verbose']:
        logger.setLevel(logging.DEBUG)

    bastion = args['bastion']

    if not bastion:
        logger.info('Running in local mode. Expecting "oc" in PATH')

    try:
        with oc.client_host(hostname=bastion, username="root", auto_add_host=True, load_system_host_keys=False), oc.timeout(60 * 60):
            with oc.tracking() as t, oc.timeout(30 * 60):
                try:
                    result = oc.invoke('get', cmd_args=['events', '--field-selector', 'type==Warning'])
                    print(result.out())
                except (ValueError, OpenShiftPythonException, Exception) as e:
                    traceback.print_exc()
                    logger.error("Tracking:\n{}\n\n".format(t.get_result().as_json(redact_streams=False)))
                    raise e

    except (ValueError, OpenShiftPythonException, Exception):
        traceback.print_exc()
        exit(1)

    exit(0)
```
